devito/finite_differences/derivative.py

```diff
@@ -277,8 +285,25 @@ def evaluate(self):
         # types of discretizations.
         return self._eval_fd(self.expr)
 
+    @property
+    def _eval_deriv(self):
+        return self._eval_fd(self.expr)
+
     def _eval_fd(self, expr):
-        expr = getattr(expr, 'evaluate', expr)
+        """
+        valuate finite difference approximation of the Derivative.
+        The evaluation goes in four steps:
+        - 1: evaluate derivatives within the expression. For example `f.dx * g` will
+        evaluate `f.dx` first.
+        - 2: Evaluate the finite difference for the (new) expression
+        - 3: Evaluate remaining (as `g` may need to be evaluated at a different point)
+        - 4: apply subsititutions.
+
+        """
+        # Evaluate derivatives in the expression
+        was_mul = expr.is_Mul
```

Comment:
nitpicking: instead of `was_mul`, you could name the output variables `expr1` , `expr2`, ... for a 1-to-1 correspondence to the four steps of the docstring

you may even separate the conceptual blocks of the method by interleaving each of the four steps with a blank line, and prefixing code with a comment along the lines of `# Step 1: ...`